src/uct/ib/dc/dc_mlx5_ep.h

```diff
@@ -592,15 +592,16 @@ static inline struct mlx5_grh_av *uct_dc_mlx5_ep_get_grh(uct_dc_mlx5_ep_t *ep)
                          (_iface)->super.super.config.fc_hard_thresh)) { \
             ucs_status_t _status = uct_dc_mlx5_ep_check_fc(_iface, _ep); \
             if (ucs_unlikely(_status != UCS_OK)) { \
-                if (((_ep)->dci != UCT_DC_MLX5_EP_NO_DCI) && \
-                    !uct_dc_mlx5_iface_is_dci_rand(_iface)) { \
-                    ucs_assertv_always(uct_dc_mlx5_iface_dci_has_outstanding(_iface, \
-                                                                             (_ep)->dci), \
-                                       "iface (%p) ep (%p) dci leak detected: dci=%d", \
-                                       _iface, _ep, (_ep)->dci); \
-                } \
                 return _status; \
             } \
+            \
+            if ((_ep)->fc.fc_wnd <= 0) { \
+                UCS_STATS_UPDATE_COUNTER((_ep)->fc.stats, \
+                                         UCT_RC_FC_STAT_NO_CRED, 1); \
+                UCS_STATS_UPDATE_COUNTER((_ep)->super.stats, \
+                                         UCT_EP_STAT_NO_RES, 1); \
+                return UCS_ERR_NO_RESOURCE; \
+            } \
```

Comment:
imo better check it inside `uct_dc_mlx5_ep_check_fc`, because it is looks strange that `uct_dc_mlx5_ep_check_fc` returns `UCS_OK`, while no FC resources are available.

Reply:
it is checked here to not increment `UCT_RC_FC_STAT_NO_CRED` and `UCT_EP_STAT_NO_RES` in DC ep_check(), because even if these resources aren't available we send do keepalive anyway

Reply:
how is `keepalive` related here?